docs/proposals/component-registration.md

```diff
@@ -0,0 +1,320 @@
+
+
+## Use Cases
+
+This proposal primarily covers features that can be achieved internally by Kubernetes itself, but many of these features and changes are motivated by external use cases, which need enablement. Not all of these use cases will be completely enabled by this proposal.
+
+
+### Deployment
+
+As a k8s operator, I want a to be able to add/remove new components (instances or types) to a new or running cluster (ideally without manually registering it with the apiserver).
+
+As a k8s operator, I want a kubectl command that allows me to validate that the cluster has been deployed and is fully functional.
+
+As a k8s operator, I want a kubectl command that allows me to check which components are deployed/installed on a specific cluster in order to infer which features it supports.
+
+Note: Inferring features from components isn't very useful with vanilla k8s, but becomes more useful as 3rd party or optional components are added (e.g. OpenShift) and/or if components-controller is split into multiple components.
+
+### Debugging
+
+As a k8s operator, I want a single endpoint (and kubectl command) where I can see the current condition of the cluster so that I can debug observed failures (dropped connections, unresponsiveness, etc.).
+
+Note: The endpoint should contain cause where possible, but at least the current condition of the components.
+
+### Notification
+
+As a k8s operator, I want to be notified when the cluster requires human intervention so that I can intervene.
+
+As an external notification system (e.g. PagerDuty), I want an endpoint that responds to regular requests with the current cluster condition so that I can determine (with my own rule set) when to notify the k8s operator.
+
+Note: What condition requires human intervention is not a rule set that can be hard-coded, is likely to be different for each cloud platform, changes based on component configuration, and is likely to require a time-series database.
+
+### Reporting
+
+As a k8s operator, I want to see a report/graph of the readiness/uptime/availability of the cluster over time so that I can make/satisfy SLAs.
+
+Note: At least requires a time-series database and probably a reporting engine.
+
+### Self-Healing
+
+As an k8s operator, I want the cluster to self-heal when individual components crash so that I don't get notified or need to intervene.
+
+As a local process monitoring tool (e.g. monit), I want to be able to determine the condition of the component process I am monitoring to determine if it needs to be restarted.
+
+As a remote container/vm monitoring tool (e.g. bosh monitor), I want to be able to determine the health of the component container/vm I am monitoring.
+
+Note: Self-healing requires knowing how to deploy the cluster components, which is different for each cloud platform and changes based on component configuration.
+
```

Comment:
> The reason why ingress is mentioned is that these components will live on the cluster, but to be visible to clients, must be exposed outside the cluster (be nodeport, load balanced service, or ingress). To a client, the method of ingress/egress, just the outcome (there is an address the client can talk to).

This confuses me, I think because of terminology, but I think we're on the same page.

The definition I put in the proposal considers the **Primary Components** as being *external* to the k8s cluster: their lifecycle is not managed by k8s; they are not deployed on k8s nodes; and they do not have (and do not have to have) a kube-proxy or equivalent running locally to route Pod or Service IPs. They may be somewhere else entirely, on different machines, with different network rules, as long as the k8s nodes (kubelet + kube-proxy) can assess the apiserver. 

The **Addon Components**, on the other hand, ARE deployed by/on k8s. 

And there's potentially a third class of components that are both optional and external to the cluster, for which we don't have a name yet. I assume the "logging component" and "build service" you mentioned belong in this third class of external optional components. If so, then they're not part of the cluster and would be registered just like Primary Components, with a Service and one or more ExternalServiceTargets. Since the ExternalServiceTargets are outside of the k8s cluster, they need only an egress address, not an ingress address. If you added an ingress in addition, it would be like telling k8s to manage things outside of its deployment scope.

ExternalServiceTargets (or "ServiceProviders", as I'm leaning towards) will be usable for more than just k8s components. As far as k8s is concerned, the external service providers only need to be accessible from inside the cluster (internal-to-external, aka egress) so that Pods can use them. But they don't necessarily need to be made accessible to ALL k8s pods, if the operator wants to set up some network partitions so that only certain namespaces can use certain services.

There is an additional complexity here tho: external components are not required to be on a host with a kube-proxy. So an external controller may not be able to connect to all ExternalServiceTargets, because the ExternalServiceTarget address is from the perspective of a Pod, for k8s cluster egress. So if a controller is expected to perform liveness/readiness probes, it would either need to be an Addon Component, deployed with k8s, or have similar networking so that egress from that component's host uses the same address as egress from a Pod. But I think that's just a networking constraint the operator will have to take into account, not something we need to codify in the API.

Anyway, if you disagree with the Discovery section I drafted, please rewrite it so I can understand then differences. Otherwise I'll merge it into the proposal.